src/pwm.rs

```diff
@@ -0,0 +1,1683 @@
+    feature = "stm32f303xd",
+    feature = "stm32f303xe",
+    feature = "stm32f358",
+    feature = "stm32f398"
+))]
+use crate::gpio::gpiof::PF10;
+#[cfg(any(
+    feature = "stm32f302",
+    feature = "stm32f303xb",
+    feature = "stm32f303xc",
+    feature = "stm32f303xd",
+    feature = "stm32f303xe",
+    feature = "stm32f373",
+    feature = "stm32f378",
+    feature = "stm32f358",
+    feature = "stm32f398"
+))]
+use crate::gpio::gpiof::PF6;
+#[cfg(any(
+    feature = "stm32f302",
+    feature = "stm32f303xb",
+    feature = "stm32f303xc",
+    feature = "stm32f303xd",
+    feature = "stm32f303xe",
+    feature = "stm32f373",
+    feature = "stm32f358",
+    feature = "stm32f398"
+))]
+use crate::gpio::gpiof::PF9;
+
+use crate::rcc::Clocks;
+use crate::stm32::RCC;
+use crate::time::Hertz;
+
+/// Output Compare Channel 1 of Timer 1 (type state)
+pub struct TIM2_CH1 {}
+/// Output Compare Channel 2 of Timer 1 (type state)
+pub struct TIM2_CH2 {}
+/// Output Compare Channel 3 of Timer 1 (type state)
+pub struct TIM2_CH3 {}
+/// Output Compare Channel 4 of Timer 1 (type state)
+pub struct TIM2_CH4 {}
+
+/// Output Compare Channel 1 of Timer 15 (type state)
+pub struct TIM15_CH1 {}
+/// Output Compare Channel 2 of Timer 15 (type state)
+pub struct TIM15_CH2 {}
+
+/// Output Compare Channel 1 of Timer 16 (type state)
+pub struct TIM16_CH1 {}
+
+/// Output Compare Channel 1 of Timer 17 (type state)
+pub struct TIM17_CH1 {}
+
+/// Type state used to represent a channel that has no pins yet
+pub struct NoPins {}
+/// Type state used to represent a channel is using regular pins
+pub struct WithPins {}
+/// Type state used to represent a channel is using (only) complementary pins
+pub struct WithNPins {}
+
+/// Representation of a Channel for an abritary timer channel,
+/// that also holds a type state for whether or not this channel
+/// is using any pins yet.
+///
+/// If there are no pins supplied, it cannot be enabled.
+pub struct PwmChannel<X, T> {
+    timx_chy: PhantomData<X>,
+    pin_status: PhantomData<T>,
+}
+
+macro_rules! pwm_timer_private {
+    ($timx:ident, $TIMx:ty, $res:ty, $apbxenr:ident, $apbxrstr:ident, $pclkz:ident, $timxrst:ident, $timxen:ident, $enable_break_timer:expr, [$($TIMx_CHy:ident),+], [$($x:ident),+]) => {
+        /// Create one or more output channels from a TIM Peripheral
+        /// This function requires the maximum resolution of the duty cycle,
+        /// the period of the PWM signal and the frozen clock configuration.
+        ///
+        /// The resolution should be chosen to offer sufficient steps against
+        /// your target peripheral.  For example, a servo that can turn from
+        /// 0 degrees (2% duty cycle) to 180 degrees (4% duty cycle) might choose
+        /// a resolution of 9000.  This allows the servo to be set in increments
+        /// of exactly one degree.
+        pub fn $timx(tim: $TIMx, res: $res, freq: Hertz, clocks: &Clocks) -> ($(PwmChannel<$TIMx_CHy, NoPins>),+) {
```

Comment:
Could we silence the warnings here?

```suggestion
        #[allow(unused_parens)]
        pub fn $timx(tim: $TIMx, res: $res, freq: Hertz, clocks: &Clocks) -> ($(PwmChannel<$TIMx_CHy, NoPins>),+) {
```

Reply:
Yeah, no simple way to avoid the warning--this is part of what makes single channels not result in a tuple.